File: src/jobhunter/scan.py
"""Job-scan domain types (spec 2026-06-26). Pure data + storage interface.

No I/O, no SQL, no FastAPI — import-safe without a DB so routes can be tested
against an in-memory fake (tests/fake_scan_store.py).
"""

import builtins
from dataclasses import asdict, dataclass
from typing import Protocol

SITES: tuple[str, ...] = ("indeed", "onlinejobs_ph", "jobstreet", "linkedin")
CANDIDATE_STATUSES: tuple[str, ...] = ("new", "generated", "dismissed")
SCAN_STATUSES: tuple[str, ...] = ("completed", "partial")
INITIAL_CANDIDATE_STATUS = "new"
PICKS_MIN, PICKS_MAX = 1, 10


def validate_site(site: str) -> None:
    if site not in SITES:
        raise ValueError(f"unknown site: {site!r} (allowed: {', '.join(SITES)})")


def validate_candidate_status(status: str) -> None:
    if status not in CANDIDATE_STATUSES:
        raise ValueError(
            f"unknown status: {status!r} (allowed: {', '.join(CANDIDATE_STATUSES)})"
        )


def validate_settings(
    search_titles: builtins.list[str],
    sites_enabled: builtins.list[str],
    picks_per_site: int,
) -> None:
    if not search_titles or any(not t.strip() for t in search_titles):
        raise ValueError("search_titles must be a non-empty list of non-blank strings")
    if not sites_enabled:
        raise ValueError("sites_enabled must not be empty")
    for s in sites_enabled:
        validate_site(s)
    if not (PICKS_MIN <= picks_per_site <= PICKS_MAX):
        raise ValueError(f"picks_per_site must be between {PICKS_MIN} and {PICKS_MAX}")


@dataclass
class ScanSettings:
    search_titles: builtins.list[str]
    sites_enabled: builtins.list[str]
    picks_per_site: int
    enabled: bool
    updated_at: str
    location: str = ""  # configurable search location (e.g. "Philippines", "Remote")

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class ScanStatus:
    status: str  # idle | running | completed | error
    started_at: str | None
    finished_at: str | None
    new_count: int
    site_summary: dict

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class CandidateInput:
    site: str
    url: str
    title: str
    company: str | None
    location: str | None
    jd_text: str
    fit_reason: str | None
    fit_score: float | None


@dataclass
class Candidate:
    id: str
    scan_id: str
    site: str
    url: str
    title: str
    company: str | None
    location: str | None
    jd_text: str
    fit_reason: str | None
    fit_score: float | None
    status: str
    slug: str | None
    created_at: str

    def to_dict(self) -> dict:
        return asdict(self)


@dataclass
class Scan:
    id: str
    started_at: str | None
    finished_at: str | None
    status: str
    site_summary: dict
    created_at: str

    def to_dict(self) -> dict:
        return asdict(self)


class ScanStore(Protocol):
    def get_settings(self) -> ScanSettings: ...

    def update_settings(
        self,
        *,
        search_titles: builtins.list[str],
        sites_enabled: builtins.list[str],
        picks_per_site: int,
        enabled: bool,
        location: str = "",
    ) -> ScanSettings: ...

    def known_urls(self) -> builtins.list[str]: ...

    def record_scan(
        self,
        *,
        started_at: str | None,
        finished_at: str | None,
        status: str,
        site_summary: dict,
        candidates: builtins.list[CandidateInput],
    ) -> tuple[Scan, int, int]:
        """Insert scan + new candidates. Returns (scan, new_count, skipped_count)."""
        ...

    def list_candidates(
        self, *, status: str | None = None, scan_id: str | None = None
    ) -> builtins.list[Candidate]: ...

    def get_candidate(self, candidate_id: str) -> Candidate | None: ...

    def set_candidate_status(
        self, candidate_id: str, *, status: str, slug: str | None = None
    ) -> Candidate | None: ...

    def list_scans(self) -> builtins.list[Scan]: ...

    def mark_scan_running(self) -> ScanStatus: ...

    def mark_scan_completed(
        self, *, new_count: int, site_summary: dict
    ) -> ScanStatus: ...

    def get_scan_status(self) -> ScanStatus: ...


__all__ = [
    "SITES", "CANDIDATE_STATUSES", "SCAN_STATUSES", "INITIAL_CANDIDATE_STATUS",
    "PICKS_MIN", "PICKS_MAX", "validate_site", "validate_candidate_status",
    "validate_settings", "ScanSettings", "ScanStatus", "CandidateInput",
    "Candidate", "Scan", "ScanStore",
]

File: src/jobhunter/web/frontend/src/api/scan.ts
export type Site = "indeed" | "onlinejobs_ph" | "jobstreet" | "linkedin";
export const SITES: Site[] = ["indeed", "onlinejobs_ph", "jobstreet", "linkedin"];
export const SITE_LABEL: Record<Site, string> = {
  indeed: "Indeed",
  onlinejobs_ph: "OnlineJobs PH",
  jobstreet: "JobStreet",
  linkedin: "LinkedIn",
};

export type ScanSettings = {
  search_titles: string[];
  sites_enabled: Site[];
  picks_per_site: number;
  enabled: boolean;
  location?: string;
  updated_at: string;
};

export type CandidateStatus = "new" | "generated" | "dismissed";

export type Candidate = {
  id: string;
  scan_id: string;
  site: Site;
  url: string;
  title: string;
  company: string | null;
  location: string | null;
  jd_text: string;
  fit_reason: string | null;
  fit_score: number | null;
  status: CandidateStatus;
  slug: string | null;
  created_at: string;
};

export type Scan = {
  id: string;
  started_at: string | null;
  finished_at: string | null;
  status: string;
  site_summary: Record<string, { status: string; count: number }>;
  created_at: string;
};

async function json<T>(resp: Response, what: string): Promise<T> {
  if (!resp.ok) throw new Error(`${what} failed: ${resp.status}`);
  return resp.json() as Promise<T>;
}

export async function getScanSettings(): Promise<ScanSettings> {
  return json(await fetch("/api/scan/settings"), "getScanSettings");
}

export async function putScanSettings(s: ScanSettings): Promise<ScanSettings> {
  return json(
    await fetch("/api/scan/settings", {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(s),
    }),
    "putScanSettings",
  );
}

export async function listScans(): Promise<Scan[]> {
  return json(await fetch("/api/scan/scans"), "listScans");
}

export async function listCandidates(scanId?: string): Promise<Candidate[]> {
  const q = scanId ? `?scan_id=${encodeURIComponent(scanId)}` : "";
  return json(await fetch(`/api/scan/candidates${q}`), "listCandidates");
}

export async function dismissCandidate(id: string): Promise<Candidate> {
  return json(
    await fetch(`/api/scan/candidates/${encodeURIComponent(id)}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ status: "dismissed" }),
    }),
    "dismissCandidate",
  );
}

export async function generateFromCandidate(
  id: string,
): Promise<{ slug: string; status: string }> {
  return json(
    await fetch(`/api/scan/candidates/${encodeURIComponent(id)}/generate`, {
      method: "POST",
    }),
    "generateFromCandidate",
  );
}

// Manually trigger a scan run via the external n8n engine. Surfaces the API's
// detail string on failure (e.g. "scan engine not configured").
export async function runScan(): Promise<{ triggered: boolean }> {
  const resp = await fetch("/api/scan/run", { method: "POST" });
  if (!resp.ok) {
    let detail = `${resp.status}`;
    try {
      const body = await resp.json();
      if (body?.detail) detail = body.detail;
    } catch {
      // non-JSON error body; keep the status code
    }
    throw new Error(detail);
  }
  return resp.json();
}
